src/async_impl/decoder.rs

```diff
@@ -0,0 +1,433 @@
+/*!
+A potentially non-blocking response decoder.
+
+The decoder wraps a stream of chunks and produces a new stream of decompressed chunks.
+The decompressed chunks aren't guaranteed to align to the compressed ones.
+
+If the response is plaintext then no additional work is carried out.
+Chunks are just passed along.
+
+If the response is gzip, then the chunks are decompressed into a buffer.
+Slices of that buffer are emitted as new chunks.
+
+This module consists of a few main types:
+
+- `ReadableChunks` is a `Read`-like wrapper around a stream
+- `Decoder` is a layer over `ReadableChunks` that applies the right decompression
+
+The following types directly support the gzip compression case:
+
+- `Pending` is a non-blocking constructor for a `Decoder` in case the body needs to be checked for EOF
+- `Peeked` is a buffer that keeps a few bytes available so `libflate`s `read_exact` calls won't fail
+*/
+
```

Comment:
The work in the module is phenomenal! It was quite easy to read and follow what was happening, thanks!

One comment is that the style throughout the rest of the crate is to prefer generics of just a single letter. As for my personal opinion, I find seeing argument types that are a word makes me immediately assume it's some concrete type, like `TStream`, instead of `T`, where I know that is a generic.